[BC-376] Support snappy over gossip

mbaxter opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/teku/blob/master/CONTRIBUTING.md -->

## PR Description
Add support for snappy compression over gossip.

Also:
* Fixed some ssz serialization issues:
  * Consider extra bytes to be an error
  * Decode plain `boolean` and `Boolean` types
  * Throw explicit error if reflection information is missing
  * Throw error if primitive type cannot be explicitly decoded

mbaxter committed:
Rework Compressor API for use with gossip

mbaxter committed:
Make SimpleOffsetSerializer more robust / stricter, add tests

mbaxter committed:
Add GossipEncoding classes

mbaxter committed:
Use new GossipEncoder's when handling gossip

mbaxter committed:
Update integration tests to run across different gossip encodings

mbaxter committed:
Merge branch 'master' into bc-376/support-snappy-over-gossip

Reconcile filename changes

mbaxter committed:
Add more roundtrip encoding tests for relevant gossip datastructures

mbaxter committed:
Fix null result handling

ajsutton commented on networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/compression/SnappyCompressor.java:
```diff
@@ -45,31 +47,60 @@ public Bytes compress(final Bytes data) {
+                uncompressedPayloadSize.getAsInt(), uncompressed.size()));
       }
-      if (snappyIn.available() > 0) {
+      if (validateSize && snappyIn.available() > 0) {
```
Shouldn't we be doing this check in all cases? ie even if we don't know what the expected uncompressed size is, we should check that we actually read all bytes from the gossip message.

mbaxter replied:
It turns out we're supposed to use different compression strategies for gossip vs rpc requests (snappy block-format vs framed-format compression).  So, reverted these changes.

ajsutton reviewed (dismissed):
LGTM. Particularly nice to make SSZ decoding more explicit about failures.

mbaxter committed:
Merge branch 'master' into bc-376/support-snappy-over-gossip

mbaxter committed:
Revert "Rework Compressor API for use with gossip"

This reverts commit f0bcb615f8188e2b7f5aba84f048a4108014b0f0.

mbaxter committed:
Use distinct compression strategies for gossip vs rpc requests

Gossip compression uses snappy block format, rpc request use the snappy
framed format.

mbaxter committed:
Add logging around topic subscription

mbaxter committed:
Fix snappy gossip encoding name

cemozerr reviewed (approved):
LGTM.
